content/docs/digging_deeper/locks.md

```diff
@@ -0,0 +1,357 @@
+
+Also, the environment variable must be validated to allow one of the pre-configured stores. The validation is defined inside the `start/env.ts` file using the `Env.schema.enum` rule.
+
+```ts
+{
+  LOCK_STORE: Env.schema.enum(['redis', 'database', 'memory'] as const),
+}
+```
+
+### Redis store
+The `redis` store has a peer dependency on the `@adonisjs/redis` package; therefore, you must configure this package before using the Redis store.
+
+Following is the list of options the Redis store accepts (alongside the shared options).
+
+```ts
+{
+  redis: stores.redis({
+    connectionName: 'main',
+  }),
+}
+```
+
+<dl>
+<dt>
+connectionName
+</dt>
+<dd>
+
+The `connectionName` property refers to a connection defined within the `config/redis.ts` file.
+
+</dd>
+</dl>
+
+### Database store
+
+The `database` store has a peer dependency on the `@adonisjs/lucid` package, and therefore, you must configure this package before using the database store.
+
+Following is the list of options the database store accepts (alongside the shared options).
+
+```ts
+{
+  database: stores.database({
+    connectionName: 'postgres',
+    tableName: 'my_locks',
+  }),
+}
+```
+
+<dl>
+
+<dt>
+
+connectionName
+
+</dt>
+
+Reference to the database connection defined within the `config/database.ts` file. If not defined, we will use the default database connection.
+
+</dd>
+
+<dt>
+
+tableName
+
+</dt>
+
+<dd>
+
+The database table to use to store rate limits. 
+
+</dd>
+
+</dl>
+
+### Memory store
+
+The `memory` store is a simple in-memory store that can be useful for testing purposes but not only. Sometimes, for some use cases, you might want to have a lock that is only valid for the current process and not shared across multiple ones.
+
+The memory store is built on top of the [`async-mutex`](https://www.npmjs.com/package/async-mutex) package.
+
+```ts
+{
+  memory: stores.memory(),
+}
+```
+
+## Locking a resource
+
+Once you have configured your lock store, you can start using locks to protect your resources anywhere within your application.
+
+Here is a simple example of how to use locks to protect a resource.
+
+
+:::codegroup
+
+```ts
+// title: Manual locking
+import { errors } from '@adonisjs/lock'
+import locks from '@adonisjs/lock/services/main'
+import { HttpContext } from '@adonisjs/core/http'
+
+export default class OrderController {
+  public async process({ response, request }: HttpContext) {
+    const orderId = request.input('order_id')
+
+    const lock = locks.createLock(`order.processing.${orderId}`)
+
+    /**
+     * Try to acquire the lock immediately. Otherwise, throw an exception
+     * if the lock is already acquired.
+     */
+    try {
+      await lock.acquireImmediately()
+
+      /**
+       * Lock has been acquired. We can process the order
+       */
+      await processOrder()
```

Comment:
What happens if `processOrder` throws exception, will the lock be released? Seems like no to me

Reply:
yeah, lock won't be released in that case. will add it in the snippet

Reply:
Cool, do you think returning a boolean from `acquireImmediately` can offer a better flow? Don't have to do it right now, but for some future refactoring.

```ts
const acquired = await lock.acquireImmediately()
if (!acquired) {
  return 'Order is already being processed'
}

try {
  await processOrder()
} finally {
 await lock.release()
}
```